lib/guardian/plug/sliding_cookie.ex

```diff
@@ -0,0 +1,115 @@
+    `sliding_cookie/3` function is not provided, in which case it is called
+    with a type of `:implementation_fault` and reason `:no_sliding_cookie_fn`.
+
+    This, like all other Guardian plugs, requires a Guardian pipeline to be setup.
+    It requires an implementation module, an error handler and a key.
+
+    These can be set either:
+
+    1. Upstream on the connection with `plug Guardian.Pipeline`
+    2. Upstream on the connection with `Guardian.Pipeline.{put_module, put_error_handler, put_key}`
+    3. Inline with an option of `:module`, `:error_handler`, `:key`
+
+    Nothing is done with the token, refreshed or not, no errors are handled as validity and expiry
+    can be checked by the VerifyCookie and EnsureAuthenticated plugs respectively.
+
+    Options:
+
+    * `:key` - The location of the token (default `:default`)
+    * `:sliding_cookie` - The time (after issue) after which a replacement will be issued. Defaults to configured values.
+    * `:halt` - Whether to halt the connection in case of error. Defaults to `true`.
+
+    The `:sliding_cookie` config (or plug option) should be the same format as `:ttl`, for example
+    `{1, :hour}`, and obviously it should be less than the prevailing `:ttl`.
+    """
+
+    import Plug.Conn
+    import Guardian.Plug.Keys
+
+    alias Guardian.Plug.Pipeline
+
+    import Guardian, only: [ttl_to_seconds: 1, decode_and_verify: 4, timestamp: 0]
+
+    @behaviour Plug
+
+    @impl Plug
+    @spec init(opts :: Keyword.t()) :: Keyword.t()
+    def init(opts), do: opts
+
+    @impl Plug
+    @spec call(conn :: Plug.Conn.t(), opts :: Keyword.t()) :: Plug.Conn.t()
+    def call(%{req_cookies: %Plug.Conn.Unfetched{}} = conn, opts) do
+      conn
+      |> fetch_cookies()
+      |> call(opts)
+    end
+
+    def call(conn, opts) do
+      with {:ok, token} <- find_token_from_cookies(conn, opts),
+           module <- Pipeline.fetch_module!(conn, opts),
+           {:ok, refresh_after} <- sliding_window(module, opts),
+           {:ok, %{"iat" => iat} = claims} <- decode_and_verify(module, token, %{}, opts),
+           {:ok, resource} <- module.resource_from_claims(claims),
+           true <- timestamp() > iat + refresh_after,
+           {:ok, new_c} <- module.sliding_cookie(claims, resource, opts) do
+        conn
+        |> Guardian.Plug.remember_me(module, resource, new_c)
+      else
+        {:error, :not_implemented} ->
+          conn
+          |> Pipeline.fetch_error_handler!(opts)
+          |> apply(:auth_error, [conn, {:implementation_fault, :no_sliding_cookie_fn}, opts])
+          |> Guardian.Plug.maybe_halt(opts)
+
+        _ ->
+          conn
+      end
+    end
+
+    defp sliding_window(module, opts) do
+      case Keyword.get(opts, :sliding_cookie, module.config(:sliding_cookie)) do
+        nil ->
+          :no_sliding_window
+
+        ttl_descr ->
+          {:ok, ttl_to_seconds(ttl_descr)}
+      end
+    end
+
+    defp find_token_from_cookies(conn, opts) do
```

Comment:
Maybe we should share this function between verifyCookie and slidingCookie?

Reply:
Sure. I don't like code duplication but let this pass as it was a small function, and I wasn't sure it was worth coupling the two modules to share it. I don't feel strongly either way though, happy to call the `VerifyCookie` version.

Reply:
Ah, could move it to plug.ex.

Reply:
Would be great I think we have some code in `plug` that basically does this aswell

Reply:
I looked at that, it has quite a lot in common with `fetch_token_key/2`, but it they don't really refactor together nicely IMO; I think the result would just be more abstract less clear code. The kicker is `find_token_from_cookies/2` does `token = conn.req_cookies[key] || conn.req_cookies[to_string(key)]`, using a string and atom version of the key. I'm not sure if this is necessary, the unit tests certainly pass fine with just token = conn.req_cookies[to_string(key)]` and my app is ok with it, but I'm a little reluctant to change such an obviously deliberate effort without more research, I don't know what behaviour older versions of Phoenix might have had that might necessitate it, or something else I haven't thought of...